src/linearmap.rs

```diff
@@ -0,0 +1,248 @@
+// Copyright 2022 The aarch64-paging Authors.
+// This project is dual-licensed under Apache 2.0 and MIT terms.
+// See LICENSE-APACHE and LICENSE-MIT for details.
+
+//! Functionality for managing page tables with linear mapping.
+
+use crate::{
+    paging::{Attributes, MemoryRegion, PhysicalAddress, RootTable, Translation, VirtualAddress},
+    AddressRangeError,
+};
+#[cfg(target_arch = "aarch64")]
+use core::arch::asm;
+
+/// Manages a level 1 page-table using linear mapping, where every virtual address is either
+/// unmapped or mapped to an IPA with a fixed offset.
+///
+/// Mappings should be added with [`map_range`](Self::map_range) before calling
+/// [`activate`](Self::activate) to start using the new page table. To make changes which may
+/// require break-before-make semantics you must first call [`deactivate`](Self::deactivate) to
+/// switch back to a previous static page table, and then `activate` again after making the desired
+/// changes.
+///
+/// # Example
+///
+/// ```
+/// use aarch64_paging::{
+///     linearmap::LinearMap,
+///     paging::{Attributes, MemoryRegion},
+/// };
+///
+/// const ASID: usize = 1;
+/// const ROOT_LEVEL: usize = 1;
+/// const OFFSET: isize = 4096;
+///
+/// // Create a new page table with linear mapping.
+/// let mut pagetable = LinearMap::new(ASID, ROOT_LEVEL, OFFSET);
+/// // Map a 2 MiB region of memory as read-write.
+/// pagetable.map_range(
+///     &MemoryRegion::new(0x80200000, 0x80400000),
+///     Attributes::NORMAL | Attributes::NON_GLOBAL | Attributes::EXECUTE_NEVER,
+/// ).unwrap();
+/// // Set `TTBR0_EL1` to activate the page table.
+/// # #[cfg(target_arch = "aarch64")]
+/// pagetable.activate();
+///
+/// // Write something to the memory...
+///
+/// // Restore `TTBR0_EL1` to its earlier value while we modify the page table.
+/// # #[cfg(target_arch = "aarch64")]
+/// pagetable.deactivate();
+/// // Now change the mapping to read-only and executable.
+/// pagetable.map_range(
+///     &MemoryRegion::new(0x80200000, 0x80400000),
+///     Attributes::NORMAL | Attributes::NON_GLOBAL | Attributes::READ_ONLY,
+/// ).unwrap();
+/// # #[cfg(target_arch = "aarch64")]
+/// pagetable.activate();
+/// ```
+#[derive(Debug)]
+pub struct LinearMap {
+    root: RootTable<LinearTranslation>,
+    #[allow(unused)]
+    asid: usize,
+    #[allow(unused)]
+    previous_ttbr: Option<usize>,
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+struct LinearTranslation {
+    /// The offset from a virtual address to the corresponding (intermediate) physical address.
+    offset: isize,
+}
+
+impl Translation for LinearTranslation {
+    fn virtual_to_physical(&self, va: VirtualAddress) -> PhysicalAddress {
+        PhysicalAddress(va.0.wrapping_add(self.offset as usize))
```

Comment:
I was thinking that wrapping keeps things simple as every virtual address always has a corresponding physical address (and vice-versa), but actually that doesn't work as neither is likely to allow the full 64-bit address range.

I've changed this to panic instead. I think it makes sense to panic for a failure to translate physical addresses to virtual, as that implies that the page table contains an invalid subtable address (which should never happen). For a failure to translate virtual addresses to physical it would in some cases be better to return an error, but the way the code currently works we'd only discover the issue partway through mapping a range, by which point we can't safely roll back. The alternative would be to do a first walk to validate the mapping then a second to update the pagetable. I'll look into doing that, but it will have a runtime overhead.

Reply:
Refactored to return an error rather than panicking. Actually it is possible to do this without a separate page table walk, as the range is contiguous so checking the first and last page is enough.